File: tools/godsplit_plan.py
#!/usr/bin/env python3
"""godsplit_plan — the boundary + hazard planner for a behavior-preserving Go split.

The ``/modularize`` skill retires the code-quality scorecard's ``architecture`` debt
(god-files > 1500 lines, god-functions > 200 lines) by MOVING top-level declarations
into concern-scoped files in the SAME package — a semantic no-op in Go EXCEPT for the
hazards this tool surfaces.

The error-prone part of a clean split is cutting at the right line: a top-level decl's
doc comment sits ABOVE its ``func``/``type`` keyword, so a naive cut at the keyword
orphans the comment. This tool computes, for every top-level declaration, the
DOC-COMMENT-AWARE block boundaries (the exact ``sed -n 'A,Bp'`` range to extract), and
flags the four things that make code motion NOT a no-op:

* per-file build tags (``//go:build``) — moving a decl changes which file carries them;
* ``func init()`` — init order is filename-ALPHABETICAL across a package, so moving an
  ``init()`` between files can silently reorder initialization;
* aliased imports (``x "path"``) — ``goimports -w`` re-derives plain imports after a move
  but does NOT re-infer a local alias, so an alias must be re-added to the new file by hand;
* multi-line backtick raw strings — this is a LINE parser, not a Go tokenizer; embedded
  column-0 ``func``/``type`` text inside a raw string is correctly SKIPPED, but the
  ``raw_strings`` count warns you to eyeball the plan for such a file.

It is line-based and best-effort: ``decl`` detection ignores raw-string interiors and the
package/import header, and func length uses brace-depth over string/comment-stripped code.
Always sanity-check (see the skill): reject any plan where a decl's ``block_end <
block_start``, or where the decl count disagrees with ``grep -cE '^(func|type|var|const) '``.

Read-only. Run from the repo root::

    python tools/godsplit_plan.py internal/model/kv.go            # human table
    python tools/godsplit_plan.py internal/model/kv.go --json     # machine payload

It never edits, moves, or commits anything — it only plans. The ``/modularize`` skill
turns the plan into the actual sed-extract + goimports + gofmt + verify + commit pass.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from typing import Any

FILE_HARD_MAX = 1500  # a Go file longer than this is a god-file (scorecard architecture debt)
FUNC_HARD_MAX = 200   # a function/method longer than this is a god-function

_DECL_RE = re.compile(r"^(func|type|var|const)\b")
_BLANK_RE = re.compile(r"^\s*$")
_PACKAGE_RE = re.compile(r"^package\s+(\w+)")
_METHOD_RE = re.compile(r"^func\s+\([^)]*\)\s+(\w+)")
_FUNC_RE = re.compile(r"^func\s+(\w+)")
_TYPEVARCONST_RE = re.compile(r"^(?:type|var|const)\s+(\w+)")
_BUILDTAG_RE = re.compile(r"^//\s*(?:go:build|\+build)\b")
_INIT_RE = re.compile(r"^func\s+init\s*\(")


def decl_name(line: str) -> tuple[str, str]:
    """Return ``(kind, name)`` for a top-level declaration line. A method
    ``func (r R) Name(`` reports kind ``method`` and name ``Name``; a grouped
    ``var (`` / ``const (`` block reports name ``(group)``."""
    if line.startswith("func "):
        m = _METHOD_RE.match(line)
        if m:
            return "method", m.group(1)
        m = _FUNC_RE.match(line)
        return "func", m.group(1) if m else "?"
    m = _TYPEVARCONST_RE.match(line)
    if m:
        kind = line.split(None, 1)[0]
        return kind, m.group(1)
    kind = line.split(None, 1)[0] if line.strip() else "?"
    return kind, "(group)"


def _scan_line(line: str, in_raw: bool) -> tuple[str, bool]:
    """Return ``(code, in_raw_after)`` where ``code`` is LINE with line-comments and the
    contents of ``"…"`` / ``'…'`` / backtick raw strings removed (braces preserved only
    when they are real code). ``in_raw`` carries a multi-line backtick string across
    lines. Best-effort: ``/* … */`` block comments are not handled (rare inside a decl)."""
    out: list[str] = []
    i, n = 0, len(line)
    while i < n:
        c = line[i]
        if in_raw:
            if c == "`":
                in_raw = False
            i += 1
            continue
        if c == "`":
            in_raw = True
            i += 1
            continue
        if c == "/" and i + 1 < n and line[i + 1] == "/":
            break  # line comment: rest is not code
        if c in ('"', "'"):
            quote = c
            i += 1
            while i < n:
                if line[i] == "\\":
                    i += 2
                    continue
                if line[i] == quote:
                    i += 1
                    break
                i += 1
            continue
        out.append(c)
        i += 1
    return "".join(out), in_raw


def _header_end(lines: list[str]) -> int:
    """1-based line below which top-level decls legitimately begin — the max of the
    ``package`` line, the closing ``)`` of an ``import ( … )`` block, and any single-line
    ``import "…"``. Used to clamp a decl's block_start so an extract can never swallow the
    package clause (which would duplicate it in the new file → build break)."""
    pkg = imp_end = 0
    in_import = False
    for i, line in enumerate(lines):
        s = line.strip()
        if s.startswith("package "):
            pkg = i + 1
        elif s == "import (":
            in_import = True
        elif in_import and s == ")":
            imp_end = i + 1
            in_import = False
        elif not in_import and s.startswith("import ") and '"' in s:
            imp_end = max(imp_end, i + 1)
    return max(pkg, imp_end)


def doc_start(lines: list[str], decl_idx: int, in_raw_before: list[bool]) -> int:
    """1-based line where DECL_IDX's block begins, INCLUDING the leading doc comment —
    the line after the last REAL (not raw-string-interior) blank line above the decl, or
    line 1 if none. ``decl_idx`` is 0-based into ``lines``."""
    last_blank = -1
    for i in range(decl_idx):
        if _BLANK_RE.match(lines[i]) and not in_raw_before[i]:
            last_blank = i
    return last_blank + 2


def func_span(stripped: list[str], decl_idx: int) -> int:
    """1-based END line of the func/method starting at ``decl_idx`` (0-based), by
    brace-depth over the string/comment-stripped code: the function ends when depth
    returns to 0 after the first ``{``. Robust to one-liners, indented composite
    literals, and braces inside strings/raw strings (which were stripped)."""
    depth = 0
    opened = False
    for j in range(decl_idx, len(stripped)):
        for ch in stripped[j]:
            if ch == "{":
                depth += 1
                opened = True
            elif ch == "}":
                depth -= 1
        if opened and depth <= 0:
            return j + 1
    return len(stripped)


def import_aliases(lines: list[str]) -> list[str]:
    """Aliased imports (``alias "path"`` where alias is a real local name, not ``_``/``.``
    and not just the path's basename) — the ones ``goimports`` will NOT re-create."""
    aliases: list[str] = []
    in_block = False
    for line in lines:
        s = line.strip()
        if s == "import (":
            in_block = True
            continue
        if in_block and s == ")":
            in_block = False
            continue
        target = s
        if not in_block:
            if s.startswith("import ") and '"' in s:
                target = s[len("import "):].strip()
            else:
                continue
        m = re.match(r'^([A-Za-z_]\w*)\s+"([^"]+)"', target)
        if not m:
            continue
        alias, path = m.group(1), m.group(2)
        if alias in ("_", "."):
            continue
        if alias != path.rsplit("/", 1)[-1]:
            aliases.append(f'{alias} "{path}"')
    return aliases


def plan(text: str) -> dict[str, Any]:
    """Pure fold: given Go source TEXT, return the split plan — package, hazards, and the
    decl list with doc-comment-aware block ranges + size flags. Decl/build-tag/init
    detection ignores raw-string interiors; block_start is clamped below the header. No
    file I/O, so the test drives it with a string."""
    lines = text.split("\n")
    if lines and lines[-1] == "":
        lines = lines[:-1]  # a trailing newline yields one phantom empty element
    n = len(lines)

    # Pass 1: strip strings/comments per line, tracking multi-line raw strings.
    stripped: list[str] = []
    in_raw_before: list[bool] = []
    raw_strings = 0
    in_raw = False
    for line in lines:
        in_raw_before.append(in_raw)
        code, after = _scan_line(line, in_raw)
        if not in_raw and after:  # a multi-line raw string just opened
            raw_strings += 1
        stripped.append(code)
        in_raw = after

    header_end = _header_end(lines)

    package = ""
    build_tags: list[str] = []
    init_funcs = 0
    decls: list[dict[str, Any]] = []

    # Pass 2: detect decls/build-tags/init only on REAL code lines (not raw interiors).
    for i, line in enumerate(lines):
        if in_raw_before[i]:
            continue
        if not package:
            pm = _PACKAGE_RE.match(line)
            if pm:
                package = pm.group(1)
        if _BUILDTAG_RE.match(line):
            build_tags.append(line.strip())
        if _INIT_RE.match(line):
            init_funcs += 1
        if _DECL_RE.match(line):
            kind, name = decl_name(line)
            start = max(doc_start(lines, i, in_raw_before), header_end + 1)
            d: dict[str, Any] = {"line": i + 1, "kind": kind, "name": name, "block_start": start}
            if kind in ("func", "method"):
                end = func_span(stripped, i)
                d["func_end"] = end
                d["func_lines"] = end - (i + 1) + 1
                d["god_function"] = d["func_lines"] > FUNC_HARD_MAX
            decls.append(d)

    for idx, d in enumerate(decls):
        d["block_end"] = (decls[idx + 1]["block_start"] - 1) if idx + 1 < len(decls) else n

    return {
        "package": package,
        "line_count": n,
        "god_file": n > FILE_HARD_MAX,
        "hazards": {
            "build_tags": build_tags,
            "init_funcs": init_funcs,
            "import_aliases": import_aliases(lines),
            "raw_strings": raw_strings,
        },
        "decls": decls,
    }


def _render(path: str, p: dict[str, Any]) -> str:
    out: list[str] = []
    flag = "  GOD-FILE (>1500)" if p["god_file"] else ""
    out.append(f"{path}: package {p['package']}  ({p['line_count']} lines){flag}")
    hz = p["hazards"]
    out.append("hazards (these make code motion NOT a no-op):")
    out.append(f"  build_tags     : {hz['build_tags'] or 'none'}")
    init_note = " (init order is filename-alpha — moving one between files can reorder)" if hz["init_funcs"] else ""
    out.append(f"  init_funcs     : {hz['init_funcs']}{init_note}")
    alias_note = " (goimports will NOT re-add these after a move — copy them by hand)" if hz["import_aliases"] else ""
    out.append(f"  import_aliases : {hz['import_aliases'] or 'none'}{alias_note}")
    raw_note = " (line parser may misread embedded code — eyeball the ranges)" if hz["raw_strings"] else ""
    out.append(f"  raw_strings    : {hz['raw_strings']}{raw_note}")
    out.append("")
    out.append("decls (block_start..block_end is the doc-comment-aware sed extract range):")
    for d in p["decls"]:
        span = ""
        if d["kind"] in ("func", "method"):
            mark = "  <<< GOD-FUNCTION (>200)" if d.get("god_function") else ""
            span = f"  [{d['func_lines']}L]{mark}"
        warn = "  !! INVERTED RANGE — do not trust this plan" if d["block_end"] < d["block_start"] else ""
        out.append(f"  {d['block_start']:>5}..{d['block_end']:<5} {d['kind']:>6} {d['name']}{span}{warn}")
    return "\n".join(out)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Plan a doc-comment-aware, hazard-checked Go split (read-only).")
    ap.add_argument("file", help="the .go file to plan a split of")
    ap.add_argument("--json", action="store_true", help="emit the machine-readable plan")
    args = ap.parse_args(argv)
    try:
        with open(args.file, "r", encoding="utf-8") as fh:
            text = fh.read()
    except OSError as e:
        print(f"godsplit_plan: {e}", file=sys.stderr)
        return 1
    p = plan(text)
    if args.json:
        print(json.dumps(p, indent=2))
    else:
        print(_render(args.file, p))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/godsplitplan/godsplitplan.go
// Package godsplitplan is the boundary + hazard planner for a behavior-preserving
// Go split — the Go port of the retired tools/godsplit_plan.py (fak pythongate).
//
// The /modularize skill retires the code-quality scorecard's `architecture` debt
// (god-files > 1500 lines, god-functions > 200 lines) by MOVING top-level
// declarations into concern-scoped files in the SAME package — a semantic no-op in
// Go EXCEPT for the hazards this planner surfaces.
//
// The error-prone part of a clean split is cutting at the right line: a top-level
// decl's doc comment sits ABOVE its func/type keyword, so a naive cut at the keyword
// orphans the comment. Plan computes, for every top-level declaration, the
// DOC-COMMENT-AWARE block boundaries (the exact block_start..block_end extract range)
// and flags the four things that make code motion NOT a no-op:
//
//   - per-file build tags (//go:build) — moving a decl changes which file carries them;
//   - func init() — init order is filename-ALPHABETICAL across a package, so moving an
//     init() between files can silently reorder initialization;
//   - aliased imports (x "path") — goimports -w re-derives plain imports after a move
//     but does NOT re-infer a local alias, so an alias must be re-added by hand;
//   - multi-line backtick raw strings — this is a LINE parser, not a Go tokenizer;
//     embedded column-0 func/type text inside a raw string is correctly SKIPPED, but
//     the raw_strings count warns you to eyeball the plan for such a file.
//
// It is line-based and best-effort: decl detection ignores raw-string interiors and
// the package/import header, and func length uses brace-depth over
// string/comment-stripped code. The whole fold is pure (Plan drives off an in-memory
// string), so it unit-tests without a repo. Read-only — it never edits, moves, or
// commits; it only plans.
package godsplitplan

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"regexp"
	"strings"
)

// FileHardMax: a Go file longer than this is a god-file (scorecard architecture debt).
// FuncHardMax: a function/method longer than this is a god-function.
const (
	FileHardMax = 1500
	FuncHardMax = 200
)

var (
	declRE         = regexp.MustCompile(`^(func|type|var|const)\b`)
	packageRE      = regexp.MustCompile(`^package\s+(\w+)`)
	methodRE       = regexp.MustCompile(`^func\s+\([^)]*\)\s+(\w+)`)
	funcRE         = regexp.MustCompile(`^func\s+(\w+)`)
	typeVarConstRE = regexp.MustCompile(`^(?:type|var|const)\s+(\w+)`)
	buildTagRE     = regexp.MustCompile(`^//\s*(?:go:build|\+build)\b`)
	initRE         = regexp.MustCompile(`^func\s+init\s*\(`)
	aliasRE        = regexp.MustCompile(`^([A-Za-z_]\w*)\s+"([^"]+)"`)
)

// Decl is one top-level declaration with its doc-comment-aware extract range. The
// func-only fields (FuncEnd/FuncLines/GodFunction) are pointers so they marshal to
// JSON only for func/method decls — matching the Python payload, which omits them
// entirely for type/var/const while still emitting god_function:false for a func.
type Decl struct {
	Line        int    `json:"line"`
	Kind        string `json:"kind"`
	Name        string `json:"name"`
	BlockStart  int    `json:"block_start"`
	FuncEnd     *int   `json:"func_end,omitempty"`
	FuncLines   *int   `json:"func_lines,omitempty"`
	GodFunction *bool  `json:"god_function,omitempty"`
	BlockEnd    int    `json:"block_end"`
}

// Hazards are the four signals that make code motion NOT a no-op.
type Hazards struct {
	BuildTags     []string `json:"build_tags"`
	InitFuncs     int      `json:"init_funcs"`
	ImportAliases []string `json:"import_aliases"`
	RawStrings    int      `json:"raw_strings"`
}

// Plan is the full split plan for one Go file.
type Plan struct {
	Package   string  `json:"package"`
	LineCount int     `json:"line_count"`
	GodFile   bool    `json:"god_file"`
	Hazards   Hazards `json:"hazards"`
	Decls     []Decl  `json:"decls"`
}

// DeclName returns (kind, name) for a top-level declaration line. A method
// `func (r R) Name(` reports kind "method" and name "Name"; a grouped `var (` /
// `const (` block reports name "(group)".
func DeclName(line string) (string, string) {
	if strings.HasPrefix(line, "func ") {
		if m := methodRE.FindStringSubmatch(line); m != nil {
			return "method", m[1]
		}
		if m := funcRE.FindStringSubmatch(line); m != nil {
			return "func", m[1]
		}
		return "func", "?"
	}
	if m := typeVarConstRE.FindStringSubmatch(line); m != nil {
		return firstToken(line), m[1]
	}
	return firstToken(line), "(group)"
}

// firstToken is Python's line.split(None, 1)[0] — the first whitespace-delimited
// token, or "?" for a blank line.
func firstToken(line string) string {
	f := strings.Fields(line)
	if len(f) == 0 {
		return "?"
	}
	return f[0]
}

// scanLine returns (code, inRawAfter) where code is line with line-comments and the
// contents of "…" / '…' / backtick raw strings removed (braces preserved only when
// they are real code). inRaw carries a multi-line backtick string across lines.
// Best-effort: /* … */ block comments are not handled (rare inside a decl). Iterates
// runes so a multi-byte escape can never desync the index (Python code-point parity).
func scanLine(line string, inRaw bool) (string, bool) {
	rs := []rune(line)
	n := len(rs)
	var out []rune
	i := 0
	for i < n {
		c := rs[i]
		if inRaw {
			if c == '`' {
				inRaw = false
			}
			i++
			continue
		}
		if c == '`' {
			inRaw = true
			i++
			continue
		}
		if c == '/' && i+1 < n && rs[i+1] == '/' {
			break // line comment: rest is not code
		}
		if c == '"' || c == '\'' {
			quote := c
			i++
			for i < n {
				if rs[i] == '\\' {
					i += 2
					continue
				}
				if rs[i] == quote {
					i++
					break
				}
				i++
			}
			continue
		}
		out = append(out, c)
		i++
	}
	return string(out), inRaw
}

// headerEnd returns the 1-based line below which top-level decls legitimately begin —
// the max of the package line, the closing `)` of an import ( … ) block, and any
// single-line import "…". Clamps a decl's block_start so an extract can never swallow
// the package clause (which would duplicate it in the new file → build break).
func headerEnd(lines []string) int {
	pkg, impEnd := 0, 0
	inImport := false
	for i, line := range lines {
		s := strings.TrimSpace(line)
		switch {
		case strings.HasPrefix(s, "package "):
			pkg = i + 1
		case s == "import (":
			inImport = true
		case inImport && s == ")":
			impEnd = i + 1
			inImport = false
		case !inImport && strings.HasPrefix(s, "import ") && strings.Contains(s, `"`):
			if i+1 > impEnd {
				impEnd = i + 1
			}
		}
	}
	if pkg > impEnd {
		return pkg
	}
	return impEnd
}

// docStart returns the 1-based line where declIdx's block begins, INCLUDING the
// leading doc comment — the line after the last REAL (not raw-string-interior) blank
// line above the decl, or line 1 if none. declIdx is 0-based into lines.
func docStart(lines []string, declIdx int, inRawBefore []bool) int {
	lastBlank := -1
	for i := 0; i < declIdx; i++ {
		if strings.TrimSpace(lines[i]) == "" && !inRawBefore[i] {
			lastBlank = i
		}
	}
	return lastBlank + 2
}

// funcSpan returns the 1-based END line of the func/method starting at declIdx
// (0-based), by brace-depth over the string/comment-stripped code: the function ends
// when depth returns to 0 after the first `{`. Robust to one-liners, indented
// composite literals, and braces inside strings (which were stripped).
func funcSpan(stripped []string, declIdx int) int {
	depth := 0
	opened := false
	for j := declIdx; j < len(stripped); j++ {
		for _, ch := range stripped[j] {
			if ch == '{' {
				depth++
				opened = true
			} else if ch == '}' {
				depth--
			}
		}
		if opened && depth <= 0 {
			return j + 1
		}
	}
	return len(stripped)
}

// ImportAliases returns the aliased imports (`alias "path"` where alias is a real
// local name, not `_`/`.` and not just the path's basename) — the ones goimports will
// NOT re-create after a move.
func ImportAliases(lines []string) []string {
	aliases := []string{}
	inBlock := false
	for _, line := range lines {
		s := strings.TrimSpace(line)
		if s == "import (" {
			inBlock = true
			continue
		}
		if inBlock && s == ")" {
			inBlock = false
			continue
		}
		target := s
		if !inBlock {
			if strings.HasPrefix(s, "import ") && strings.Contains(s, `"`) {
				target = strings.TrimSpace(s[len("import "):])
			} else {
				continue
			}
		}
		m := aliasRE.FindStringSubmatch(target)
		if m == nil {
			continue
		}
		alias, path := m[1], m[2]
		if alias == "_" || alias == "." {
			continue
		}
		base := path
		if idx := strings.LastIndex(path, "/"); idx >= 0 {
			base = path[idx+1:]
		}
		if alias != base {
			aliases = append(aliases, fmt.Sprintf(`%s "%s"`, alias, path))
		}
	}
	return aliases
}

// Compute folds Go source text into the split plan — package, hazards, and the decl
// list with doc-comment-aware block ranges + size flags. Decl/build-tag/init
// detection ignores raw-string interiors; block_start is clamped below the header. No
// file I/O, so a test drives it with a string.
func Compute(text string) Plan {
	lines := strings.Split(text, "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1] // a trailing newline yields one phantom empty element
	}
	n := len(lines)

	// Pass 1: strip strings/comments per line, tracking multi-line raw strings.
	stripped := make([]string, 0, n)
	inRawBefore := make([]bool, 0, n)
	rawStrings := 0
	inRaw := false
	for _, line := range lines {
		inRawBefore = append(inRawBefore, inRaw)
		code, after := scanLine(line, inRaw)
		if !inRaw && after { // a multi-line raw string just opened
			rawStrings++
		}
		stripped = append(stripped, code)
		inRaw = after
	}

	headerEndVal := headerEnd(lines)

	pkg := ""
	buildTags := []string{}
	initFuncs := 0
	decls := []Decl{}

	// Pass 2: detect decls/build-tags/init only on REAL code lines (not raw interiors).
	for i, line := range lines {
		if inRawBefore[i] {
			continue
		}
		if pkg == "" {
			if m := packageRE.FindStringSubmatch(line); m != nil {
				pkg = m[1]
			}
		}
		if buildTagRE.MatchString(line) {
			buildTags = append(buildTags, strings.TrimSpace(line))
		}
		if initRE.MatchString(line) {
			initFuncs++
		}
		if declRE.MatchString(line) {
			kind, name := DeclName(line)
			start := docStart(lines, i, inRawBefore)
			if headerEndVal+1 > start {
				start = headerEndVal + 1
			}
			d := Decl{Line: i + 1, Kind: kind, Name: name, BlockStart: start}
			if kind == "func" || kind == "method" {
				end := funcSpan(stripped, i)
				fl := end - (i + 1) + 1
				god := fl > FuncHardMax
				d.FuncEnd = &end
				d.FuncLines = &fl
				d.GodFunction = &god
			}
			decls = append(decls, d)
		}
	}

	for idx := range decls {
		if idx+1 < len(decls) {
			decls[idx].BlockEnd = decls[idx+1].BlockStart - 1
		} else {
			decls[idx].BlockEnd = n
		}
	}

	return Plan{
		Package:   pkg,
		LineCount: n,
		GodFile:   n > FileHardMax,
		Hazards: Hazards{
			BuildTags:     buildTags,
			InitFuncs:     initFuncs,
			ImportAliases: ImportAliases(lines),
			RawStrings:    rawStrings,
		},
		Decls: decls,
	}
}

// Render is the human table (the non-JSON output). It mirrors the Python layout; the
// machine contract is the JSON, so exact byte-parity of this table is not required.
func Render(path string, p Plan) string {
	var out []string
	flag := ""
	if p.GodFile {
		flag = "  GOD-FILE (>1500)"
	}
	out = append(out, fmt.Sprintf("%s: package %s  (%d lines)%s", path, p.Package, p.LineCount, flag))
	hz := p.Hazards
	out = append(out, "hazards (these make code motion NOT a no-op):")
	out = append(out, "  build_tags     : "+orNone(hz.BuildTags))
	initNote := ""
	if hz.InitFuncs > 0 {
		initNote = " (init order is filename-alpha — moving one between files can reorder)"
	}
	out = append(out, fmt.Sprintf("  init_funcs     : %d%s", hz.InitFuncs, initNote))
	aliasNote := ""
	if len(hz.ImportAliases) > 0 {
		aliasNote = " (goimports will NOT re-add these after a move — copy them by hand)"
	}
	out = append(out, "  import_aliases : "+orNone(hz.ImportAliases)+aliasNote)
	rawNote := ""
	if hz.RawStrings > 0 {
		rawNote = " (line parser may misread embedded code — eyeball the ranges)"
	}
	out = append(out, fmt.Sprintf("  raw_strings    : %d%s", hz.RawStrings, rawNote))
	out = append(out, "")
	out = append(out, "decls (block_start..block_end is the doc-comment-aware sed extract range):")
	for _, d := range p.Decls {
		span := ""
		if d.Kind == "func" || d.Kind == "method" {
			mark := ""
			if d.GodFunction != nil && *d.GodFunction {
				mark = "  <<< GOD-FUNCTION (>200)"
			}
			fl := 0
			if d.FuncLines != nil {
				fl = *d.FuncLines
			}
			span = fmt.Sprintf("  [%dL]%s", fl, mark)
		}
		warn := ""
		if d.BlockEnd < d.BlockStart {
			warn = "  !! INVERTED RANGE — do not trust this plan"
		}
		out = append(out, fmt.Sprintf("  %5d..%-5d %6s %s%s%s", d.BlockStart, d.BlockEnd, d.Kind, d.Name, span, warn))
	}
	return strings.Join(out, "\n")
}

func orNone(xs []string) string {
	if len(xs) == 0 {
		return "none"
	}
	return strings.Join(xs, ", ")
}

// Run is the CLI entry point: `godsplit-plan <file.go> [--json]`. Exit 0 on success,
// 1 on a usage error or an unreadable file — mirroring the Python main.
func Run(stdout, stderr io.Writer, argv []string) int {
	asJSON := false
	file := ""
	for _, a := range argv {
		switch {
		case a == "--json":
			asJSON = true
		case strings.HasPrefix(a, "-"):
			// ignore unknown flags (lenient vs argparse)
		default:
			if file == "" {
				file = a
			}
		}
	}
	if file == "" {
		fmt.Fprintln(stderr, "godsplit-plan: usage: godsplit-plan <file.go> [--json]")
		return 1
	}
	data, err := os.ReadFile(file)
	if err != nil {
		fmt.Fprintf(stderr, "godsplit-plan: %v\n", err)
		return 1
	}
	p := Compute(string(data))
	if asJSON {
		b, _ := json.MarshalIndent(p, "", "  ")
		fmt.Fprintln(stdout, string(b))
	} else {
		fmt.Fprintln(stdout, Render(file, p))
	}
	return 0
}
